language/internal/cc/parser/parser.go

```diff
@@ -107,37 +137,470 @@ func tokenizer(data []byte, atEOF bool) (advance int, token []byte, err error) {
+	// Collect all tokens until end of line for easier processing of directive
+	// Can collect more then 1 line if ending with '\' character
+	ts := tokensStream{}
+	tr := p.tr
+collect:
+	for {
+		token, ok := p.tr.nextInternal(true)
+		if !ok {
+			return nil, fmt.Errorf("expected more tokens: %v", tr.scanner.Err())
+		}
+		switch token {
+		case "\\":
+			// Multiline expression, continue parsing next line
+			if next, ok := tr.peek(); ok && next == EOL {
+				_, _ = tr.next() // consume EOL
 				continue
 			}
+		case EOL:
+			// End of single line expression
+			break collect
+		default:
+			ts.tokens = append(ts.tokens, token)
 		}
 	}
-	return sourceInfo
+	parser := exprParser{ts: &ts}
+	return parser.parseOr()
+}
+
+// Parser for expressions working on already loaded and cleaned up list of tokens collected until end of possibly multine macro expression
+// Used to parse the #if <expr> conditions, handles binary (&&, ||) and unary negation (!) operators
+type exprParser struct {
+	ts *tokensStream
+}
+
+func (ep *exprParser) parseOr() (Expr, error) {
+	ts := ep.ts
+	left, err := ep.parseAnd()
+	if err != nil {
+		return nil, err
+	}
+	for ts.peek("||") {
+		_ = ts.consume("||")
+		right, err := ep.parseAnd()
+		if err != nil {
+			return nil, err
+		}
+		left = Or{left, right}
+	}
+	return left, nil
+}
+
+func (ep *exprParser) parseAnd() (Expr, error) {
+	ts := ep.ts
+	left, err := ep.parseUnary()
+	if err != nil {
+		return nil, err
+	}
+	for ts.peek("&&") {
+		_ = ts.consume("&&")
+		right, err := ep.parseUnary()
+		if err != nil {
+			return nil, err
+		}
+		left = And{left, right}
+	}
+	return left, nil
+}
+
+func (ep *exprParser) parseUnary() (Expr, error) {
+	ts := ep.ts
+	switch {
+	case ts.peek("!"):
+		_ = ts.consume("!")
+		expr, err := ep.parseUnary()
+		if err != nil {
+			return nil, err
+		}
+		return Not{expr}, nil
+
+	case ts.peek("("):
+		_ = ts.consume("(")
+		expr, err := ep.parseOr()
+		if err != nil {
+			return nil, err
+		}
+		if err := ts.consume(")"); err != nil {
+			return nil, err
+		}
+		return expr, err
+
+	case ts.peek("defined"):
+		_ = ts.consume("defined")
+		if ts.peek("(") {
+			_ = ts.consume("(")
+			name := Ident(ts.next())
+			if err := ts.consume(")"); err != nil {
+				return nil, err
+			}
+			return Defined{Name: name}, nil
+		}
+		return Defined{Name: Ident(ts.next())}, nil
+	}
+
+	token := ts.next()
+	if ts.idx < len(ts.tokens) && isBinaryCompareOperator(ts.tokens[ts.idx]) {
+		op := ts.next() // ==, !=, <, ...
+		lValue, err := interpretValue(token)
+		if err != nil {
+			return nil, err
+		}
+		rightToken := ts.next()
+		rValue, err := interpretValue(rightToken)
+		if err != nil {
+			return nil, err
+		}
+		return Compare{Left: lValue, Op: op, Right: rValue}, nil
+	}
+	return Compare{Left: Ident(token), Op: "!=", Right: Constant(0)}, nil
+}
+
+// interpretValue converts a token into either Ident or Constant.
+func interpretValue(token string) (Value, error) {
```

Comment:
Maybe `parseValue` instead, as a method so it's more uniform? This isn't really interpreting.